lib/Differentiator/ForwardModeVisitor.cpp

```diff
@@ -1018,4 +1018,120 @@ namespace clad {
+    }
+    Expr* condClone = WS->getCond() ? Clone(WS->getCond()) : nullptr;
+
+    Sema::ConditionResult condRes;
+    if (condVarClone) {
+      condRes = m_Sema.ActOnConditionVariable(condVarClone, noLoc,
+                                              Sema::ConditionKind::Boolean);
+    } else {
+      condRes = m_Sema.ActOnCondition(getCurrentScope(), noLoc, condClone,
+                                      Sema::ConditionKind::Boolean);
+    }
+
+    const Stmt* body = WS->getBody();
+    Stmt* bodyResult = nullptr;
+    if (isa<CompoundStmt>(body)) {
+      bodyResult = Visit(body).getStmt();
+    } else {
+      beginScope(Scope::DeclScope);
+      beginBlock();
+      StmtDiff Result = Visit(body);
+      for (Stmt* S : Result.getBothStmts())
+        addToCurrentBlock(S);
+      CompoundStmt* Block = endBlock();
+      endScope();
+      bodyResult = Block;
+    }
+    // Since condition variable is created and initialized at each iteration,
+    // derivative of condition variable should also get created and initialized
+    // at each iteratrion. Therefore, we need to insert declaration statement
+    // of derivative of condition variable, if any, on top of the derived body
+    // of the while loop.
+    //
+    // while (double b = a) {
+    //   ...
+    //   ...
+    // }
+    //
+    // gets differentiated to,
+    //
+    // while (double b = a) {
+    //   _d_b = _d_a;
+    //   ...
+    //   ...
+    // }
+    if (condVarClone) {
+      bodyResult =
+          InsertAtFrontOfCompoundStmt(m_Sema.getASTContext(),
+                                      cast<CompoundStmt>(bodyResult),
+                                      BuildDeclStmt(condVarRes.getDecl_dx()));
+    }
+
+    auto WSDiff = clad_compat::Sema_ActOnWhileStmt(m_Sema, condRes, bodyResult)
+                      .getAs<WhileStmt>();
+    // end scope for while loop
+    endScope();
+    return StmtDiff(WSDiff);
+  }
+
+  StmtDiff ForwardModeVisitor::VisitContinueStmt(const ContinueStmt* ContStmt) {
+    return StmtDiff(Clone(ContStmt));
+  }
+
+  StmtDiff ForwardModeVisitor::VisitDoStmt(const DoStmt* DS) {
+    // begin scope for do-while statement
+    beginScope(Scope::ContinueScope | Scope::BreakScope);
+    Expr* clonedCond = DS->getCond() ? Clone(DS->getCond()) : nullptr;
+    const Stmt* body = DS->getBody();
+
+    Stmt* bodyResult = nullptr;
+    if (isa<CompoundStmt>(body)) {
+      bodyResult = Visit(body).getStmt();
+    } else {
+      beginScope(Scope::DeclScope);
+      beginBlock();
+      StmtDiff Result = Visit(body);
+      for (Stmt* S : Result.getBothStmts())
+        addToCurrentBlock(S);
+      CompoundStmt* Block = endBlock();
+      endScope();
+      bodyResult = Block;
+    }
+
+    DoStmt* DoStmtDiff =
+        m_Sema
+            .ActOnDoStmt(/*DoLoc=*/noLoc, bodyResult, /*WhileLoc=*/noLoc,
+                         /*CondLParen=*/noLoc, clonedCond, /*CondRParen=*/noLoc)
+            .getAs<DoStmt>();
```

Comment:
```suggestion
    Stmt* S =
        m_Sema
            .ActOnDoStmt(/*DoLoc=*/noLoc, bodyResult, /*WhileLoc=*/noLoc,
                         /*CondLParen=*/noLoc, clonedCond, /*CondRParen=*/noLoc)
            .get();
```
We do not need the concrete type of the statement for the return result of the function.

Reply:
Okay